lib/AST/TypeJoinMeet.cpp

```diff
@@ -17,100 +17,197 @@
-    if (auto *metatype2 = type2->getAs<ExistentialMetatypeType>()) {
-      auto instance1 = metatype1->getInstanceType();
-      auto instance2 = metatype2->getInstanceType();
-      auto result = Type::join(instance1, instance2);
-      if (!result)
-        return result;
-      return ExistentialMetatypeType::get(result);
-    }
+  static Type getSuperclassJoin(Type first, Type second);
+
+  Type visitClassType(Type second);
+  Type visitBoundGenericClassType(Type second);
+  Type visitArchetypeType(Type second);
+  Type visitDynamicSelfType(Type second);
+  Type visitMetatypeType(Type second);
+  Type visitExistentialMetatypeType(Type second);
+  Type visitBoundGenericEnumType(Type second);
+
+  Type visitOptionalType(Type second);
+
+  Type visitType(Type second) {
+    // FIXME: Implement all the visitors.
+    //    llvm_unreachable("Unimplemented type visitor!");
+    // return First->getASTContext().TheAnyType;
+    return nullptr;
   }
 
-  // If both are class types or opaque types that potentially have superclasses,
-  // find the common superclass.
-  if (type1->mayHaveSuperclass() && type2->mayHaveSuperclass()) {
-    /// Walk the superclasses of type1 looking for type2. Record them for our
-    /// second step.
-    llvm::SmallPtrSet<CanType, 8> superclassesOfType1;
-    CanType canType2 = type2->getCanonicalType();
-    for (Type super1 = type1; super1; super1 = super1->getSuperclass()) {
-      CanType canSuper1 = super1->getCanonicalType();
+public:
+  static Type join(Type first, Type second) {
+    if (!first || !second) {
+      if (first)
+        return ErrorType::get(first->getASTContext());
 
-      // If we have found the second type, we're done.
-      if (canSuper1 == canType2) return super1;
+      if (second)
+        return ErrorType::get(second->getASTContext());
 
-      superclassesOfType1.insert(canSuper1);
+      return Type();
     }
 
-    // Look through the superclasses of type2 to determine if any were also
-    // superclasses of type1.
-    for (Type super2 = type2; super2; super2 = super2->getSuperclass()) {
-      CanType canSuper2 = super2->getCanonicalType();
+    assert(!first->hasTypeVariable() && !second->hasTypeVariable() &&
+           "Cannot compute join of types involving type variables");
 
-      // If we found the first type, we're done.
-      if (superclassesOfType1.count(canSuper2)) return super2;
-    }
+    assert(first->getWithoutSpecifierType()->isEqual(first) &&
+           "Expected simple type!");
+    assert(second->getWithoutSpecifierType()->isEqual(second) &&
+           "Expected simple type!");
+
+    // If the types are equivalent, the join is obvious.
+    if (first->isEqual(second))
+      return first;
 
-    // There is no common superclass; we're done.
+    // Until we handle all the combinations of joins, we need to make
+    // sure we visit the optional side.
+    OptionalTypeKind otk;
+    if (second->getAnyOptionalObjectType(otk))
```

Comment:
Hmmm, this is a little odd. Perhaps you want to only do this "swap" when `second` is optional but `first` is not?

Reply:
Or maybe when `second` is "more optional" than `first`, because we can have multiple levels of optional and we want the left-hand side to have the larger stack of optionals.

Reply:
Yeah this is totally a hack to make the `visit()` dispatch to the `BoundGenericEnum` case to handle the optional. As things stand we support exactly what the old `join` supported. I need to get the more general levels-of-optionality and `IUO` behavior working too.